components/suggest/src/db.rs

```diff
@@ -417,6 +470,120 @@ impl<'a> SuggestDao<'a> {
         Ok(())
     }
 
+    /// Inserts all suggestions from a downloaded Pocket attachment into
+    /// the database.
+    pub fn insert_pocket_suggestions(
+        &mut self,
+        record_id: &SuggestRecordId,
+        suggestions: &[DownloadedPocketSuggestion],
+    ) -> Result<()> {
+        for suggestion in suggestions {
+            self.scope.err_if_interrupted()?;
+            let suggestion_id: i64 = self.conn.query_row_and_then_cachable(
+                "INSERT INTO suggestions(
+                     record_id,
+                     provider,
+                     title,
+                     url
+                 )
+                 VALUES(
+                     :record_id,
+                     :provider,
+                     :title,
+                     :url
+                 )
+                 RETURNING id
+                 ",
+                named_params! {
+                    ":record_id": record_id.as_str(),
+                    ":provider": SuggestionProvider::Pocket,
+                    ":title": suggestion.title,
+                    ":url": suggestion.url,
+
+                },
+                |row| row.get(0),
+                true,
+            )?;
+            self.conn.execute(
+                "INSERT INTO pocket_custom_details(
+                             suggestion_id,
+                             description,
+                             score
+                         )
+                         VALUES(
+                             :suggestion_id,
+                             :description,
+                             :score
+                         )",
+                named_params! {
+                    ":suggestion_id": suggestion_id,
+                    ":description": suggestion.description,
+                    ":score": suggestion.score,
+                },
+            )?;
+            for (_, keyword) in suggestion.low_confidence_keywords.iter().enumerate() {
+                let (keyword_prefix, keyword_suffix) = match keyword.find(' ') {
+                    Some(index) => {
+                        let (prefix, suffix) = keyword.split_at(index);
+                        (prefix.to_string(), suffix[1..].to_string())
+                    }
+                    None => (keyword.clone(), String::new()),
+                };
```

Comment:
Let's lean on the Rust standard library here!

```suggestion
                let (keyword_prefix, keyword_suffix) = keyword.split_once(' ').unwrap_or_else(|| (keyword, ""));
```

`split_once` will do the checking and slicing for you, and returns references, so we can also avoid the extra copies with `to_string()` and `clone()`.

Reply:
These loops also look mostly identical! We can use some of Rust's iterator methods to cut down on the repetition:

```rust
for ((rank, keyword), confidence) in suggestion
    .high_confidence_keywords
    .iter()
    .enumerate()
    .zip(std::iter::repeat(PocketKeywordConfidence::High))
    .chain(
        suggestion
            .low_confidence_keywords
            .iter()
            .enumerate()
            .zip(std::iter::repeat(PocketKeywordConfidence::Low)),
    )
{
    // ...
}
```

`std::iter::repeat` returns an infinite iterator, and `zip`-ping another iterator with it will combine elements from both of them, and stop as soon as one of the iterators is exhausted.

Reply:
😵 wow this is fancy